assets/__tests__/rendered_test.js

```diff
@@ -0,0 +1,191 @@
+import { Rendered } from '../js/phoenix_live_view'
+
+describe('Rendered', () => {
+  describe('mergeDiff', () => {
+    it('recursively merges two diffs', () => {
+      expect(Rendered.mergeDiff(simpleDiff1, simpleDiff2)).toEqual(
+        simpleDiffResult
+      )
+      expect(Rendered.mergeDiff(deepDiff1, deepDiff2)).toEqual(deepDiffResult)
+    })
+    it('returns the latter diff if it contains a `static` key', () => {
+      const diff1 = { 0: ['a'], 1: ['b'] }
+      const diff2 = { 0: ['c'], static: 'c' }
+      expect(Rendered.mergeDiff(diff1, diff2)).toEqual(diff2)
+    })
+  })
+
+  describe('isNewFingerprint', () => {
+    it('returns true if `diff.static` is truthy', () => {
+      const diff = { static: ['<h2>'] }
+      expect(Rendered.isNewFingerprint(diff)).toEqual(true)
+    })
+    it('returns false if `diff.static` is falsy', () => {
+      const diff = { static: undefined }
+      expect(Rendered.isNewFingerprint(diff)).toEqual(false)
+    })
+    it('returns false if `diff` is undefined', () => {
+      expect(Rendered.isNewFingerprint()).toEqual(false)
+    })
+  })
+
+  describe('toString', () => {
+    it('stringifies a diff', () => {
+      expect(Rendered.toString(simpleDiffResult).trim()).toEqual(
+        `
+<div class="thermostat">
+  <div class="bar cooling">
+    <a href="#" phx-click="toggle-mode">cooling</a>
+    <span>07:15:04 PM</span>
+  </div>
+  <div class="controls">
+    <span class="reading">72</span>
+    <button phx-click="dec" class="minus">-</button>
+    <button phx-click="inc" class="plus">+</button>
+    <span class="weather">
+      <div conn id="phx-dafeTrsp/m0=:DemoWeb.WeatherLive"
+     data-phx-parent-id="phx-dafeTrsp/m0="
+     data-phx-view="DemoWeb.WeatherLive"
+     data-phx-session="STUBBED"></div>
+<div class="phx-loader"></div>
+
+    </span>
+  </div>
+</div>
+      `.trim()
+      )
+    })
+  })
+
+  describe('toOutputBuffer', () => {})
+
+  describe('comprehensionToBuffer', () => {})
+
+  describe('dynamicToBuffer', () => {})
+})
+
+const simpleDiff1 = {
+  '0': 'cooling',
+  '1': 'cooling',
+  '2': '07:15:03 PM',
+  '3': '72',
+  '4':
+    '<div conn id="phx-dafeTrsp/m0=:DemoWeb.WeatherLive"\n     data-phx-parent-id="phx-dafeTrsp/m0="\n     data-phx-view="DemoWeb.WeatherLive"\n     data-phx-session="STUBBED"></div>\n<div class="phx-loader"></div>\n',
+  static: [
+    '<div class="thermostat">\n  <div class="bar ',
+    '">\n    <a href="#" phx-click="toggle-mode">',
+    '</a>\n    <span>',
+    '</span>\n  </div>\n  <div class="controls">\n    <span class="reading">',
+    '</span>\n    <button phx-click="dec" class="minus">-</button>\n    <button phx-click="inc" class="plus">+</button>\n    <span class="weather">\n      ',
+    '\n    </span>\n  </div>\n</div>\n',
+  ],
+}
+
+const simpleDiff2 = {
+  '2': '07:15:04 PM',
+}
+
+const simpleDiffResult = {
+  '0': 'cooling',
+  '1': 'cooling',
+  '2': '07:15:04 PM',
+  '3': '72',
+  '4':
```

Comment:
Do you think # 4 could be removed if it doesn't add anything to the test?

Reply:
Yeah, I can trim this down. 

All these examples were direct copy/paste from a Live View socket message. I didn't want to break anything by making assumptions about the data structure. But now that we have the tests in place, we can trim it down to focus more on what's being tested.